docbrown/core/src/props.rs

```diff
@@ -1,166 +1,333 @@
+                vertex_slot.set(prop_id, Some(prop));
+            }
+        }
+        Ok(())
+    }
 
-            match self.edge_meta.get_mut(edge_id) {
-                Some(edge_props) => edge_props.set(prop_id, t, prop),
-                None => self
-                    .edge_meta
-                    .insert(edge_id, TPropVec::from(prop_id, t, prop)),
+    pub fn set_static_edge_props(&mut self, edge_id: usize, props: &Vec<(String, Prop)>) -> Result<(), ()> {
+        Self::assert_valid_edge_id(edge_id);
+        if !props.is_empty() {
+            let translated_props = self.translate_props(props, true);
+            let edge_slot: &mut LazyVec<Option<Prop>> = Self::grow_and_get_slot(&mut self.static_edge_props, edge_id);
+            for (prop_id, prop) in translated_props {
+                edge_slot.set(prop_id, Some(prop));
             }
         }
+        Ok(())
+    }
+
+    fn assert_valid_edge_id(edge_id: usize) { // TODO: this should return a result
+        if edge_id == 0 {
+            panic!("Edge id (= 0) in invalid because it cannot be used to express both remote and local edges")
+        };
     }
 }
 
 #[cfg(test)]
 mod props_tests {
     use super::*;
 
-    #[test]
-    fn zero_index_of_edge_meta_is_preassgined_default_value() {
-        let Props {
-            prop_ids: _,
-            vertex_meta: _,
-            edge_meta,
-        } = Props::default();
-
-        assert_eq!(edge_meta, vec![TPropVec::Empty]);
-    }
-
     #[test]
     fn return_valid_next_available_edge_id() {
-        let props = Props::default();
+        let mut props = Props::default();
 
         // 0th index is not a valid edge id because it can't be used to correctly denote
         // both local as well as remote edge id. Hence edge ids must always start with 1.
-        assert_ne!(props.get_next_available_edge_id(), 0);
         assert_eq!(props.get_next_available_edge_id(), 1);
     }
 
     #[test]
     #[should_panic]
     fn assigning_edge_id_as_0_should_fail() {
         let mut props = Props::default();
-        props.upsert_edge_props(1, 0, &vec![]);
+        props.upsert_temporal_edge_props(1, 0, &vec![]);
     }
 
     #[test]
     fn return_prop_id_if_prop_name_found() {
         let mut props = Props::default();
-        props.prop_ids.insert(String::from("key1"), 0);
-        props.prop_ids.insert(String::from("key2"), 1);
+        props.prop_ids.insert(String::from("key1"), PropId::Temporal(0));
+        props.prop_ids.insert(String::from("key2"), PropId::Temporal(1));
 
-        assert_eq!(props.get_prop_id("key2"), 1);
+        assert_eq!(props.get_or_allocate_id("key2", false), Ok(1));
     }
 
     #[test]
     fn return_new_prop_id_if_prop_name_not_found() {
         let mut props = Props::default();
-        assert_eq!(props.get_prop_id("key1"), 0);
-        assert_eq!(props.get_prop_id("key2"), 1);
+        assert_eq!(props.get_or_allocate_id("key1", false), Ok(0));
```

Comment:
what happens if we don't use that prop id? and can we get into that situation?

Reply:
this is how Props assigns ids. The first time the user calls e.g. `add_vertex` with some prop, the name doesn't exist so we need to allocate one. By design, if this gets called is because we are gonna store some property right after that (this is all inside Props implementation)